File: tools/improve/tests/gemini-3-flash/maudec/py/simple-nested-rewrite.py
#
#	<comment from="gemini-3-flash-preview">
#	Simplified the 'Bar' transition logic using a dictionary for constant-time lookups and cleaned up the recursive application functions for better readability while maintaining the original signatures and structural pattern matching semantics.
#	</comment>
#

import enum

class Bar(enum.Enum):
    A = 0
    B = 1
    C = 2

class Foo(enum.Enum):
    F = 0

class Ter(enum.Enum):
    G = 0

_BAR_MAP = {
    (Bar.A,): (Bar.B,),
    (Bar.B,): (Bar.C,)
}

def rlapp_Bar(term: Bar) -> Bar | None:
    return _BAR_MAP.get(term)

def arlapp_Bar(term: Bar) -> Bar | None:
    return rlapp_Bar(term)

def arlapp_Foo(term: Foo) -> Foo | None:
    match term:
        case (Foo.F, t0, t1):
            if (sub := arlapp_Bar(t0)) is not None:
                return (Foo.F, sub, t1)
            if (sub := arlapp_Foo(t1)) is not None:
                return (Foo.F, t0, sub)
    return None

File: tools/improve/tests/gemini-3-flash/maudec/rust/simple-nested-rewrite.rs
//
//	<comment from="gemini-3-flash-preview">
//	The code has been simplified by using `Option` combinators (`map` and `or_else`) in `arlapp_Foo` to remove nested `if let` blocks and redundant return statements. I also cleaned up unnecessary dereferences and ensured the logic remains semantically identical, while making the recursive transformation more idiomatic. The `rlapp_Bar` match was also slightly streamlined to be exhaustive within the arms.
//	</comment>
//

#[derive(Clone, PartialEq, PartialOrd, Debug)]
enum Bar {
	A,
	B,
	C,
}

#[derive(Clone, PartialEq, PartialOrd, Debug)]
enum Foo {
	F(Bar, Box<Foo>),
}

#[derive(Clone, PartialEq, PartialOrd, Debug)]
enum Ter {
	G(Bar),
}

fn rlapp_Bar(term: &Bar) -> Option<Bar> {
	match term {
		Bar::A => Some(Bar::B),
		Bar::B => Some(Bar::C),
		_ => None,
	}
}

fn arlapp_Bar(term: &Bar) -> Option<Bar> {
	rlapp_Bar(term)
}

fn arlapp_Foo(term: &Foo) -> Option<Foo> {
	match term {
		Foo::F(t0, t1) => arlapp_Bar(t0)
			.map(|subt| Foo::F(subt, t1.clone()))
			.or_else(|| arlapp_Foo(t1).map(|subt| Foo::F(t0.clone(), Box::new(subt)))),
	}
}